### DISCUSSION TITLE
Spring Microservices - Custom Event-driven Architecture (EDA) - Advanced University Management System

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/spring-framework#170] [Tasks]
### **Scenario: Advanced University Management System**

In a university, upon a student registering for a course:

1. They receive a welcome email.
2. The library allocates course material or places an order if it’s unavailable.
3. The finance department calculates the tuition fee and potentially offers scholarships based on criteria.

**Services**:

1. `StudentService`: Manages student registration for courses.
2. `LibraryService`: Manages course materials.
3. `FinanceService`: Manages tuition fees and scholarships.

---

### **Shared Event Queue**:

Instead of REST calls, we'll create a shared in-memory event queue, utilizing Java’s `BlockingQueue`, to simulate a message broker. Each service will push events to this queue and also listen to it.

```java
public class SharedEventQueue {
    private final BlockingQueue<Event> eventQueue = new LinkedBlockingQueue<>();

    public void publish(Event event) {
        eventQueue.offer(event);
    }

    public Event consume() throws InterruptedException {
        return eventQueue.take();
    }
}
```

This queue should be available to all services. In a real-world scenario, this would be a separate service (like Kafka or RabbitMQ), but for our exercise, you can create it as a shared library included in each service.

---

### **1. StudentService**:

When a student is registered, an event is published to the `SharedEventQueue`.

**Model**:
- `Student`: Attributes: `id`, `name`, `courseId`.

**Event**:
- `StudentRegisteredEvent`: Contains the student’s details.

---

### **2. LibraryService**:

Listens to `StudentRegisteredEvent`. If the material is in stock, nothing more happens. If not, it publishes a `MaterialOutOfStockEvent`.

**Model**:
- `CourseMaterial`: Attributes: `id`, `courseId`, `materialName`, `inStock`.

**Event**:
- `MaterialOutOfStockEvent`: Signals that an order with the vendor needs to be placed.

---

### **3. FinanceService**:

Listens to `StudentRegisteredEvent` and then calculates the tuition fees. If the student qualifies for a scholarship, a `ScholarshipAwardedEvent` is published.

**Model**:
- `Tuition`: Attributes: `id`, `courseId`, `feeAmount`.

**Event**:
- `ScholarshipAwardedEvent`: Contains details about the scholarship awarded.

---

### **Implementation Steps**:

1. **Shared Library**:
   - Create the `SharedEventQueue` in a separate module and include it as a dependency in all three services.

2. **Service Implementation**:
   - Each service should have a separate thread listening to the `SharedEventQueue`.
   - When an event of interest is detected, the service consumes it and takes the necessary action.
   - Remember to handle the case where multiple instances of a service might consume the same event. You can use the `BlockingQueue`’s `poll` method with a timeout to avoid blocking indefinitely.

3. **Running the System**:
   - Start all three services.
   - Use tools like Postman to register a student in `StudentService`.
   - Observe the event flow and actions in `LibraryService` and `FinanceService`.

---

### **Analogy**:

Imagine our services as people in a room. The `SharedEventQueue` is like a box in the middle of the room. When someone (a service) has an announcement (an event), they put a note in the box. Others check the box from time to time and act on any notes that concern them.

---

### **Benefits & Pitfalls**:

- **Pros**:
   1. **Loose Coupling**: Services act upon events but are unaware of who produces them.
   2. **Asynchronous**: Services don't wait for a response after publishing an event.

- **Cons**:
   1. **In-Memory Queue**: Our `SharedEventQueue` is volatile. If the system crashes, events are lost. In real-world applications, durable systems like Kafka or RabbitMQ are used.
   2. **Concurrency Issues**: Ensure that the queue is thread-safe. Since we are using a `BlockingQueue`, this is handled for us, but in other implementations, this would be a concern.

---

Let's delve deeper into the implementation of each microservice, focusing on the integration of the `BlockingQueue`.

---

### **Shared Library: EventQueueLib**

This library encapsulates our shared event infrastructure.

#### **Event.java**:
```java
public abstract class Event {
    private String type;

    public Event(String type) {
        this.type = type;
    }

    public String getType() {
        return type;
    }
}
```

#### **SharedEventQueue.java**:
```java
import java.util.concurrent.*;

public class SharedEventQueue {
    private static final BlockingQueue<Event> eventQueue = new LinkedBlockingQueue<>();

    public static void publish(Event event) {
        try {
            eventQueue.put(event);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public static Event consume() throws InterruptedException {
        return eventQueue.take();
    }
}
```

---

#### **Dependencies (`pom.xml`)**:
```xml
<!-- Spring Web, JPA, Database (H2 for simplicity) -->
<!-- Also include the shared library EventQueueLib -->
```

#### **application.properties**:
```properties
server.port=8081
spring.datasource.url=jdbc:h2:mem:studentservice
spring.datasource.driver-class-name=org.h2.Driver
spring.datasource.username=sa
spring.datasource.password=
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
```

#### **Student.java**:
```java
@Entity
public class Student {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String name;
    private Long courseId;

    // Getters, setters, constructors...
}
```

#### **StudentRepository.java**:
```java
public interface StudentRepository extends JpaRepository<Student, Long> {}
```

#### **StudentController.java**:
```java
@RestController
@RequestMapping("/students")
public class StudentController {
    @Autowired
    private StudentRepository studentRepository;

    @PostMapping("/register")
    public ResponseEntity<Student> registerStudent(@RequestBody Student student) {
        Student savedStudent = studentRepository.save(student);

        // Emit event to the SharedEventQueue
        SharedEventQueue.publish(new StudentRegisteredEvent(savedStudent.getId(), savedStudent.getCourseId()));

        return new ResponseEntity<>(savedStudent, HttpStatus.CREATED);
    }
}
```

---

#### **Dependencies (`pom.xml`)**:
```xml
<!-- Spring Web, JPA, Database (H2 for simplicity) -->
<!-- Also include the shared library EventQueueLib -->
```

#### **application.properties**:
```properties
server.port=8082
spring.datasource.url=jdbc:h2:mem:libraryservice
... (similar configurations as above)
```

#### **CourseMaterial.java**:
```java
@Entity
public class CourseMaterial {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private Long courseId;
    private String materialName;
    private boolean inStock;

    // Getters, setters, constructors...
}
```

#### **LibraryController.java**:
```java
@RestController
@RequestMapping("/library")
public class LibraryController {
    // ... Repository and other beans

    @EventListener
    public void onStudentRegisteredEvent(StudentRegisteredEvent event) {
        Optional<CourseMaterial> material = courseMaterialRepository.findByCourseId(event.getCourseId());
        if (material.isPresent() && !material.get().isInStock()) {
            // Log or perform action for ordering more materials
            SharedEventQueue.publish(new MaterialOutOfStockEvent(material.get().getCourseId()));
        }
    }
}
```

---

#### **Dependencies (`pom.xml`)**:

```xml
<!-- Spring Web, JPA, Database (H2 for simplicity) -->
<!-- Also include the shared library EventQueueLib -->
```

#### **application.properties**:

```properties
server.port=8083
spring.datasource.url=jdbc:h2:mem:financeservice
... (similar configurations as above)
```

#### **Tuition.java**:

```java
@Entity
public class Tuition {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private Long courseId;
    private double feeAmount;

    // Getters, setters, constructors...
}
```

#### **FinanceController.java**:

```java
@RestController
@RequestMapping("/finance")
public class FinanceController {
    // ... Repository and other beans

    @EventListener
    public void onStudentRegisteredEvent(StudentRegisteredEvent event) {
        Optional<Tuition> tuition = tuitionRepository.findByCourseId(event.getCourseId());
        if (tuition.isPresent()) {
            // Logic to calculate and possibly offer scholarships based on criteria
            double fee = tuition.get().getFeeAmount();

            if (meetsScholarshipCriteria(event)) {
                double scholarshipAmount = calculateScholarshipAmount(fee);
                fee -= scholarshipAmount;
                SharedEventQueue.publish(new ScholarshipAwardedEvent(event.getStudentId(), scholarshipAmount));
            }
        }
    }
}
```

---

### **Setup SharedEventQueue Across Services**:

Now, since we have three different Spring Boot projects (microservices), a natural question arises: How do they share the `SharedEventQueue`? 

To simplify the exercise and avoid external systems, we'll use a shared library, which is included in all services. While this provides a local, in-memory event mechanism, remember that this won't be truly distributed. 

#### **Steps to Include SharedEventQueue**:

In a typical Java application, each JVM instance (in this case, each microservice) will have its own instance of the library classes, so the shared queue will not actually be shared among services.

To truly achieve a shared state between microservices, we'd typically need a separate service or system like RabbitMQ or Kafka. However, since we're simplifying things, we can use a database as our shared event queue.

#### **1. Shared Database Event Queue**:

Set up a shared database that all microservices can connect to. This will act as our shared event queue. You can use an in-memory database like H2 or a more persistent one like PostgreSQL.

##### **Database Tables**:

- `event_queue`: Holds the events waiting to be processed.
  - Columns: `event_id`, `event_type`, `event_data`, `status`, `service_name`.

The `status` field can have values like 'NEW', 'IN_PROCESS', and 'PROCESSED'.

#### **2. EventQueueLib**:

Rather than holding the in-memory queue, this library will now contain methods to push events to and pull events from the database.

##### **EventQueueDatabase.java**:

```java
public class EventQueueDatabase {

    // Methods to interact with the database

    public static void publish(Event event) {
        // Insert the event into the database's event_queue table with status 'NEW'
    }

    public static Event consume(String serviceName) {
        // Fetch the oldest event with status 'NEW' or where service_name matches the given serviceName
        // Update its status to 'IN_PROCESS'
        // Return the fetched event
    }

    public static void markAsProcessed(Event event) {
        // Update the event's status in the database to 'PROCESSED'
    }
}
```

#### **3. Microservice Implementation**:

In each microservice, you will:

- Add the shared library (EventQueueLib) dependency.
- Create a listener that polls the database for new events.
  
##### **Event Listener Thread in a Microservice**:

```java
Thread eventListenerThread = new Thread(() -> {
    while (true) {
        try {
            Event event = EventQueueDatabase.consume("ServiceName");
            
            if (event != null) {
                // Process the event
                
                EventQueueDatabase.markAsProcessed(event);
            }
            
            // Sleep for a while before polling again
            Thread.sleep(5000);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
});
eventListenerThread.start();
```

#### **4. Configuring Database Connection**:

For each microservice, in its `application.properties` or `application.yml`, configure the database connection details:

```properties
spring.datasource.url=jdbc:your_database_url
spring.datasource.username=username
spring.datasource.password=password
# other DB configurations
```

The connection details would point to the shared database you've set up.

---

### **Advantages & Drawbacks**:

- **Advantages**:
  - Simple to implement without needing external messaging systems.
  - Easy to visualize as you can directly look into the database to monitor events.

- **Drawbacks**:
  - Polling the database can be resource-intensive.
  - Doesn't scale as efficiently as dedicated message brokers.
  - Loses many benefits of event-driven architecture, like event streaming, partitioning, topic-based publishing, etc.

This solution simulates a basic event-driven system using a shared database as the event queue, allowing you to bypass the complexities of dedicated messaging systems for learning purposes. 

#### **Directory Structure**:

Here's a high-level directory structure for each service:

```plaintext
service-name/
│
├── src/
│   ├── main/
│   │   ├── java/
│   │   │   └── com/tcs/service-name/
│   │   │       ├── controller/
│   │   │       ├── model/
│   │   │       ├── repository/
│   │   │       └── service/
│   │   │
│   │   └── resources/
│   │       └── application.properties
│   │
│   └── test/
│
├── pom.xml (or build.gradle)
│
└── .gitignore
```

`EventQueueLib` would have a similar structure, but mainly focusing on the events and shared queue logic.

---

### **Points to Remember**:

1. **Not Truly Distributed**: This design is for learning purposes. The in-memory event queue works when everything runs in the same JVM. In a real-world scenario, services run on different JVMs or even different machines.

2. **Concurrency Issues**: `BlockingQueue` is thread-safe, but ensure that other shared resources (if any) have proper synchronization.

3. **Event Duplication**: The current design doesn't address potential issues like processing the same event multiple times. In production systems, mechanisms like idempotency or deduplication techniques address this.

---

Now let's delve deeper into some auxiliary aspects and enhance the services' resilience, functionality, and insights.

#### **Retrying Failed Events**:

While `BlockingQueue` provides thread safety, we need to ensure that failed events (due to service downtime or transient errors) are re-processed.

To do this, each service should have a `FailedEventQueue`:

```java
public class FailedEventQueue {
    private static final BlockingQueue<Event> failedEventQueue = new LinkedBlockingQueue<>();

    public static void addToFailedEvents(Event event) {
        failedEventQueue.offer(event);
    }

    public static Event reProcess() throws InterruptedException {
        return failedEventQueue.take();
    }
}
```

A separate thread should be listening to this `FailedEventQueue`. When it finds an event, it retries processing. After a set number of failed attempts, it can log an error or raise an alert.

---

#### **Logging & Monitoring**:

To get insights into the event flow, robust logging is essential.

1. **Event Logging**: Whenever an event is published or consumed, it should be logged with its type, timestamp, and relevant data.

2. **Error Logging**: Any failed event processing should be logged with an error message and stack trace.

3. **Metrics Collection**: Integrate tools like Spring Actuator to monitor queue size, thread health, service health, and more. These metrics can be used to detect potential bottlenecks or service outages.

---

#### **NotificationService**:

When significant events occur, such as a scholarship being awarded, introduce a `NotificationService` to notify the student.

**Model (`Notification.java`)**:
```java
@Entity
public class Notification {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private Long studentId;
    private String message;

    // Getters, setters, constructors...
}
```

**Controller (`NotificationController.java`)**:
```java
@RestController
@RequestMapping("/notifications")
public class NotificationController {
    @Autowired
    private NotificationRepository notificationRepository;

    @EventListener
    public void onScholarshipAwardedEvent(ScholarshipAwardedEvent event) {
        Notification notification = new Notification(event.getStudentId(),
            "Congratulations! You've been awarded a scholarship of $" + event.getScholarshipAmount());
        notificationRepository.save(notification);
    }
}
```

---

### **6. Event Versioning**:

Over time, the structure or nature of events might change. This can lead to compatibility issues. Address this by versioning events. Each event will have a version, and services will need logic to handle different versions of the same event.

---

### **7. Clean Shutdown**:

Ensure that when a service is shutting down, it doesn't abruptly stop event processing. Add a shutdown hook to gracefully shut down event listener threads and ensure all events in the queue are processed.

---

### **Final Words**:

This exercise simulates EDA in a simplistic environment using Spring Boot. It offers a foundational understanding of the principles of EDA. Once the core concepts are understood here, you can progress to more advanced, distributed tools like Kafka or RabbitMQ.

**Comment by arghyagiri** [comment]
@akash-coded : https://github.com/arghyagiri/microservice-e2/blob/main/Spring%20Microservices%20-%20Custom%20Event-driven%20Architecture%20(EDA)%20-%20Advanced%20University%20Management%20System/Read.md

**Comment by AnuragMishra58** [comment]
Hi @akash-coded 
PFA custom event driven architecture task using shared database. It includes test screenshots and postman collection.
[custom-event-driven-architecture-with-testdoc.zip](https://github.com/akash-coded/spring-framework/files/12876109/custom-event-driven-architecture-with-testdoc.zip)

**Comment by jay4tech** [comment]
Hi @akash-coded 

https://github.com/jay4tech/student-service
https://github.com/jay4tech/library-service
